### PR TITLE
Make Operations Picklable

### PR DESCRIPTION
**Pull request by b-butler**
## Description

This PR implements the necessary changes to make `hoomd.Operations` and `hoomd._HOOMDBaseObject` objects picklable. Overall there are many individual changes that needed to be made, but I tried to keep the commits modular, and if necessary I can move this into multiple PRs.

Note due to the limitations of pickling, if a user uses lambdas or certain file type objects in their operations, they will not be picklable. 

## Motivation and context

We want to make operations picklable for restartability and enabling simulation metadata storing.

Resolves #???

## How has this been tested?

Tests for all existing tested operations have been added.

## Change log

```
* Added: `hoomd.md.Force`, `hoomd.md.Operation`, and `hoomd.md.Operations` objects are now picklable.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by vyasr** [comment]
A feature to consider for way down the road is allowing users to provide custom pickle functions (e.g. using `dill` or `cloudpickle`) if they want to pickle something that the default `pickle` library doesn't support.

**Comment by joaander** [comment]
Some of the tests need to be updated to remove `seed`: https://dev.azure.com/glotzerlab/hoomd-blue/_build/results?buildId=10531&view=logs&j=c21dfe3d-6dd2-5969-8cf5-3e196ca72a91&t=ee49db16-ba03-55b0-4481-6e0b297f3809&l=24

I would recommend merging master up into this branch.

File: hoomd/data/typeparam.py
Comment (bdice) [inline hoomd/data/typeparam.py:5]: I can't recall the details, but there are a lot of potential problems with `__slots__` if class inheritance is involved. I believe that using `__slots__` for filtering class attributes in `__getattr__` wasn't an option in signac's synced collections for that reason. There are some notes here: https://docs.python.org/3/reference/datamodel.html#notes-on-using-slots

File: hoomd/data/typeparam.py
Comment (bdice) [inline hoomd/data/typeparam.py:64]: Typo? (Missing test?)
```suggestion
            state['param_dict'] = self.param_dict.to_detached()
```

File: hoomd/hpmc/pytest/test_quick_compress.py
Comment (bdice) [inline hoomd/hpmc/pytest/test_quick_compress.py:177]: Test comment seems inaccurate.

File: hoomd/hpmc/pytest/test_clusters.py
Comment (bdice) [inline hoomd/hpmc/pytest/test_clusters.py:231]: Test comment seems inaccurate.

File: hoomd/hpmc/tune/move_size.py
Comment (bdice) [inline hoomd/hpmc/tune/move_size.py:243]: Is this `return` a mistake?
```suggestion
```

File: hoomd/md/integrate.py
Comment (bdice) [inline hoomd/md/integrate.py:11]: Is this import used? It's newly added but I don't see where it it utilized.

File: hoomd/operation.py
Comment (bdice) [inline hoomd/operation.py:245]: An alternative would be to define something like:

```python
class _Instantiator:
    def __init__(self, instance_class):
        self.instance_class = instance_class
    def __call__(self):
        return self.instance_class()

_dependents = _Instantiator(list)
```

Or even better yet, just provide a class that will create a new instance when called. 😉 
```
_dependents = list
```

Then you could just call the item to get a value if it's callable and return values like `None` directly.

```python
# Replace whatever calls _ReturnCopies with:
return value() if callable(value) else value
```

File: hoomd/operations.py
Comment (bdice) [inline hoomd/operations.py:310]:
```suggestion
        # ensure that top level changes to self.__dict__ are not propagated
```

File: hoomd/pytest/test_table.py
Comment (bdice) [inline hoomd/pytest/test_table.py:172]: Some linters will complain about this syntax and prefer that the output value is never assigned (because it's not used). Assignment to a variable cannot have side-effects as far as I am aware, so the test behavior of raising an error should be equivalent.
```suggestion
        hoomd.write.Table(1, logger, output)
```

File: hoomd/simulation.py
Comment (bdice) [inline hoomd/simulation.py:245]: Suggestion on the comment text. But more importantly, I don't know if this gives a developer enough info to understand why an error would occur. I haven't looked at `update_group_dof` but if it's not obvious, that would be good to describe.
```suggestion
                # The _schedule method must be called after setting _operations so that
                # self.state.update_group_dof runs without errors.
```

File: hoomd/util.py
Comment (bdice) [inline hoomd/util.py:43]: `io.TextIOBase, io.BufferedIOBase, io.RawIOBase` are already subclasses of `io.IOBase`.
```suggestion
        (str, dict, io.IOBase))
```

**Review by bdice — CHANGES_REQUESTED** [review]
Comments below. I probably won't have time to get back for another round of feedback on the entire PR, so feel free to merge this without waiting for a second round of review from me. Of course you can ping me in the comments if you need further guidance on anything specific that I mentioned. 👍

> 
> 
> A feature to consider for way down the road is allowing users to provide custom pickle functions (e.g. using `dill` or `cloudpickle`) if they want to pickle something that the default `pickle` library doesn't support.

I wondered about this as well -- there are a lot of private "partial" classes that had to be introduced for this behavior to work. If you think that's detrimental to readability or maintenance, I think it is fine to require users to install cloudpickle.

File: hoomd/operation.py
Comment (b-butler) [reply hoomd/operation.py:245]: Good thought, the last suggestion is way less over engineered.

File: hoomd/md/integrate.py
Comment (b-butler) [reply hoomd/md/integrate.py:11]: This is a remnant of an intermediate commit.

File: hoomd/hpmc/tune/move_size.py
Comment (b-butler) [reply hoomd/hpmc/tune/move_size.py:243]: Yes, I do not know how that happened, maybe a botched copy and paste. Good catch.

File: hoomd/hpmc/pytest/test_quick_compress.py
Comment (b-butler) [reply hoomd/hpmc/pytest/test_quick_compress.py:177]: Good catch.

File: hoomd/data/typeparam.py
Comment (b-butler) [reply hoomd/data/typeparam.py:64]: Typo, a consistent one.

File: hoomd/data/typeparam.py
Comment (b-butler) [reply hoomd/data/typeparam.py:5]: Yes, for classes that will likely be inherited slots are not a good option. This is not the case here though.

File: hoomd/data/syncedlist.py
Comment (b-butler) [reply hoomd/data/syncedlist.py:6]: They can but I need to use partial for the second positional only argument which I do not think is doable.

File: hoomd/data/syncedlist.py
Comment (b-butler) [reply hoomd/data/syncedlist.py:54]: For minimal changes, I will add an argument to the constructor `callable_class` which when set to `True` defaults to `False` will not wrap the class, instead treating it as a callable.

File: hoomd/simulation.py
Comment (joaander) [inline hoomd/simulation.py:250]: Not for this PR, but we still need to implement hooks to call `update_group_dof` as needed to keep things up to date after attaching.

**Review by joaander — COMMENTED** [review]
This all appears fine to me. I have some high level questions which we could discuss at the next meeting.

One of my questions is about version compatibility. Say we add a new type parameter to an operation in 3.1. For API compatibility, this parameter would have a default value consistent with the behavior in 3.0. Will 3.1 be able to read a pickled dump of this operation written by 3.0?

**Comment by b-butler** [comment]
@joaander Yes and no. We would need to modify the `__setstate__` function, but otherwise adding to the spec would be fine. We just need to check if the dict in the pickle has the new key.

File: hoomd/hpmc/tune/move_size.py
Comment (bdice) [inline hoomd/hpmc/tune/move_size.py:244]:
```suggestion
                f"target {target} should be between 0 and 1.")
```

File: hoomd/data/syncedlist.py
Comment (bdice) [inline hoomd/data/syncedlist.py:52]: It appears that the default value is an identity function?

File: hoomd/md/pair/pair.py
Comment (bdice) [inline hoomd/md/pair/pair.py:1656]: The changes to documentation of potential parameters seem unrelated to the purpose of this PR. This PR is very large and I would advocate for smaller PRs in the future, because I can't review this in one pass.

File: hoomd/md/pytest/test_potential.py
Comment (bdice) [inline hoomd/md/pytest/test_potential.py:1003]:
```suggestion
```

File: hoomd/operation.py
Comment (bdice) [inline hoomd/operation.py:103]: I'm not sure if this needs to be a `set` -- are dictionary key views treated as ordered or unordered? Also probably depends on the implementation of `keys()` for this object.

File: hoomd/operation.py
Comment (bdice) [inline hoomd/operation.py:216]: Wondering if `other._skip_for_equality` matters?

File: hoomd/operation.py
Comment (bdice) [inline hoomd/operation.py:219]: This can just be a single-line `elif`.

File: hoomd/simulation.py
Comment (bdice) [inline hoomd/simulation.py:244]:
```suggestion
                # This _schedule method must be called after setting
```

File: hoomd/variant.py
Comment (bdice) [inline hoomd/variant.py:77]: These `__eq__` methods are highly redundant. I wonder if you can do something less redundant, by defining the relevant attributes in each subclass and only implementing this `__eq__` once in the parent class?

File: hoomd/write/table.py
Comment (bdice) [inline hoomd/write/table.py:42]: Indicate that this is a "check" function in its name.
```suggestion
def _ensure_writable(fh):
```

File: hoomd/write/table.py
Comment (bdice) [inline hoomd/write/table.py:285]: This is confusing. So objects using `stdout` are picklable, because that value is mapped to `None`. Are objects using other outputs (file outputs, stderr, etc.) not picklable?

**Comment by b-butler** [comment]
@bdice sorry I know it is long. :sweat_smile: Most of these changes are directly relevant to picklability, though. I could have done many smaller PRs, but I originally did not split PR's because I thought this would get merged faster than a bunch of smaller PR's given some difficulties we have had in getting a second reviews for some Python based hoomd PR's. Given the lifetime of this one though, I don't think that panned out. I do agree though in general smaller atomic PR's are better, sorry for this behemoth.

File: hoomd/write/table.py
Comment (b-butler) [reply hoomd/write/table.py:285]: No there are not currently. It may be implementable to some extent, but I didn't attempt to do so here.

File: hoomd/custom/custom_operation.py
Comment (b-butler) [reply hoomd/custom/custom_operation.py:59]: This is only called if `_action` isn't an attribute by the nature of `__getattr__`. This is to prevent infinite recursion in some edge cases where `_action` is not set.

File: hoomd/md/pair/pair.py
Comment (b-butler) [reply hoomd/md/pair/pair.py:1656]: Yes this one is unrelated. I ended up fixing it in this PR because I only noticed the error due to other changes I was making.

File: hoomd/operation.py
Comment (b-butler) [reply hoomd/operation.py:103]: A set would be more reliable.

File: hoomd/operation.py
Comment (b-butler) [reply hoomd/operation.py:216]: It shouldn't since this is set at the class level, and objects of different classes shouldn't be equal in the first place.

File: hoomd/data/syncedlist.py
Comment (b-butler) [reply hoomd/data/syncedlist.py:41]: This is from when it was a private class

**Comment by b-butler** [comment]
@bdice, I addressed all your comments. If you don't want to review again given the size of the PR, can you remove yourself from the reviewer pull, so the PR does not say changes requested.

**Review by joaander — COMMENTED** [review]
Does this PR add any user documentation that indicates what HOOMD objects are pickleable?

I added a note to #979 to add dev docs later, but we do need to document it at the reference level for the user now. 

We can later add a full fledged tutorial with restartable jobs in #736 . I've considered, but haven't decided if it is a good idea to introduce pickling in the early introductory tutorials. It would reduce the code length of the tutorials but may be too abstract for new users. What do you think?

**Comment by bdice** [comment]
> I've considered, but haven't decided if it is a good idea to introduce pickling in the early introductory tutorials. It would reduce the code length of the tutorials but may be too abstract for new users. What do you think?

For me, this decision depends on how widely pickling is supported, because it sets the user’s expectation that pickling is (more or less) universally allowed. I’m not really sure what objects don’t support pickling but if that list is short, I’d say it is fine to use in a tutorial.

**Comment by joaander** [comment]
> > I've considered, but haven't decided if it is a good idea to introduce pickling in the early introductory tutorials. It would reduce the code length of the tutorials but may be too abstract for new users. What do you think?
> 
> For me, this decision depends on how widely pickling is supported, because it sets the user’s expectation that pickling is (more or less) universally allowed. I’m not really sure what objects don’t support pickling but if that list is short, I’d say it is fine to use in a tutorial.

@b-butler correct me if I'm wrong, but this PR essentially makes all (or almost all) `Operations` objects pickleable. We could use it to replace the repeated instantiations of the HPMC integrator in the tutorial, for example: https://hoomd-blue.readthedocs.io/en/latest/tutorial/00-Introducing-HOOMD-blue/06-Equilibrating-the-System.html#Equilibration

**Comment by b-butler** [comment]
@bdice @joaander The PR does make basically everything picklable that can be. The only exception that I am aware of is the `Table` writer when using a filehandle. We could even allow this by changing it to take in a filename and mode instead of a file handle.

**Comment by b-butler** [comment]
@joander I added a basic architecture file regarding pickling in `hoomd/architecture.rst`.

File: hoomd/architecture.rst
Comment (joaander) [inline hoomd/architecture.rst:1]: Why are we adding this in `ARCHITECTURE.rst`? Is there something wrong with the existing `ARCHITECTURE.md` file? I would prefer not to maintain two separate files with similar goals.

**Review by joaander — CHANGES_REQUESTED** [review]
Please resolve the merge conflicts and provide only one `ARCHITECTURE.*` file. The only reason to use `.rst` over `.md` is if we plan to also include it in the sphinx generated docs.

**Comment by joaander** [comment]
> Edit: I did not add it to the documentation as I had forgotten what we decided on about that and its very incomplete status.

It should be mentioned at least in the base class: `hoomd.operation.Operation`.

File: hoomd/architecture.rst
Comment (b-butler) [reply hoomd/architecture.rst:1]: Honestly I somehow just missed it, I will fix that immediately.

File: hoomd/operation.py
Comment (joaander) [inline hoomd/operation.py:310]: I think you may have misunderstood my comment. I don't think there should be a link to `ARCHITECTURE.md` here, but there should be a brief description that all `Operation` instances are picklable. Without this mention in the documentation, the change log is the only place that users will be able to see pickling mentioned.

Comment (b-butler) [reply hoomd/operation.py:310]: I did misunderstand... I will fix that.

**Comment by vyasr** [comment]
This question isn't 100% relevant to this PR and I may have missed when this decision was made, but is the plan going forward for all of the meatadata/restartability features of HOOMD to be handled via pickling? I ask because of the many portability concerns associated with pickle that also make it an unsuitable format for long-term archival or sharing of such restartability information. The original JSON-based format would have avoided this problem, but at the cost of having to develop a suitable schema. Was this tradeoff deemed acceptable?

**Comment by joaander** [comment]
> This question isn't 100% relevant to this PR and I may have missed when this decision was made, but is the plan going forward for all of the meatadata/restartability features of HOOMD to be handled via pickling?

Not exactly. 

We decided to drop work on the JSON format because it was not able to handle many of the nuances our object schema needs (tuples vs lists, object references). Also, we could not identify any other suitable format other than pickle.

However, I wouldn't say that the pickle format enabled in this PR is intended to be the solution to long term archival or sharing for the reasons you state. The best format for long term archival and sharing is still the Python script itself, in combination with metadata stored by tools like signac. I see the pickle file as a temporary storage and convenience for users with multiple workflow stages. For example, one workflow stage could use the Python API to create a list of operations and save it to a pickle file. Later stages in the workflow could read that pickle file instead of duplicating code.

Users could also avoid code duplication by putting the operation generating code into a reusable function, but there are cases like the original use-case for the JSON files (mbuild/foyer) where the code that generates the force field may not be easy to install on a cluster. With pickle files, the user could run that workflow step on their workstation, transfer the pickle files to the cluster, and continue the later stages (provided that they write the pickle files in a form compatible with both systems).

**Comment by joaander** [comment]
@b-butler tests are failing after merging from master.

**Comment by vyasr** [comment]
@joaander got it. I think that makes sense and is all that we should try to support, but probably needs to be clarified for the MoSDeF people. I know you're working on the mbuild->HOOMD conversion, is the intention to create any non-temporary intermediate files, or is it just meant for including HOOMD and mbuild in the same script? In the latter case it's completely fine, but in the former we probably need to document the limitations.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**
